keps/sig-network/3866-nftables-proxy/README.md

```diff
@@ -0,0 +1,1615 @@
+
+  - Some new functionality only works correctly if the administrator
+    passes certain command-line options to kube-proxy (eg,
+    `--cluster-cidr`), but we cannot make those options be mandatory,
+    since that would break old clusters that aren't passing them.
+
+A new kube-proxy mode, which existing users would have to explicitly opt
+into, could revisit these and other decisions.
+
+### We will hopefully be able to trade 2 supported backends for 1
+
+Right now SIG Network is supporting both the `iptables` and `ipvs`
+backends of kube-proxy, and does not feel like it can ditch `ipvs`
+because of perceived performance issues with `iptables`. If we create a new
+backend which is as functional and non-buggy as `iptables` but as
+performant as `ipvs`, then we could (eventually) deprecate both of the
+existing backends and only have one Linux backend to support in the future.
+
+### Writing a new kube-proxy mode will help to focus our cleanup/refactoring efforts
+
+There is a desire to provide a "kube-proxy library" that third parties
+could use as a base for external service proxy implementations
+([KEP-3786]). The existing "core kube-proxy" code, while functional,
+is not very well designed and is not something we would want to
+support other people using in its current form.
+
+Writing a new proxy backend will force us to look over all of this
+shared code again, and perhaps give us new ideas on how it can be
+cleaned up, rationalized, and optimized.
+
+[KEP-3786]: https://github.com/kubernetes/enhancements/issues/3786
+
+### Goals
+
+- Design and implement an `nftables` mode for kube-proxy.
+
+    - Consider various fixes to legacy `iptables` mode behavior.
+
+        - Do not enable the `route_localnet` sysctl.
+
+        - Add a more restrictive startup mode to kube-proxy, which
+          will error out if the configuration is invalid (e.g.,
+          "`--detect-local-mode ClusterCIDR`" without specifying
+          "`--cluster-cidr`") or incomplete (e.g.,
+          partially-dual-stack but not fully-dual-stack).
+
+        - (Possibly other changes discussed in this KEP.)
+
+    - Document specific details of the nftables implementation that we
+      want to consider as "API". (eg, documenting the `priority` of
+      certain chains so that administrators or other components can
+      reliably create other chains that act either before or after
+      kube-proxy's.)
+
+- Document incompatible changes between `iptables` mode and `nftables`
+  mode (e.g. localhost NodePorts, firewall handling, etc).
+
+- Do some sort of performance testing comparing the `iptables`,
+  `ipvs`, and `nftables` backends in small, medium, and large
+  clusters, comparing both the "control plane" aspects (time/CPU usage
+  spent reprogramming rules) and "data plane" aspects (latency and
+  throughput of packets to service IPs).
+
+- Come up with at least a vague plan to eventually make `nftables` the
```

Comment:
Is it actually needed in this KEP? I thought that early on it was clearly stated that it'll be a subject of entirely new KEP.  Or maybe at least point out that it's optional goal.